src/server/pkg/storage/chunk/writer.go

```diff
@@ -403,70 +468,138 @@ func (w *Writer) ChunkCount() int64 {
+	a.buf.Write(data)
+	w.bufSize += len(data)
+	w.stats.taggedBytesSize += int64(len(data))
 	written += len(data)
-	w.stats.annotatedBytesSize += int64(written)
 	return written, nil
 }
 
-func (w *Writer) writeByteSet() {
+func (w *Writer) writeDataSet() {
+	w.finishTag()
 	prev := w.prev
 	next := &chanSet{
-		bytes: make(chan *byteSet, 1),
-		done:  make(chan struct{}),
-	}
-	byteSet := &byteSet{
-		data:        w.buf.Bytes(),
-		annotations: w.annotations,
+		dataSet: make(chan *dataSet, 1),
+		done:    make(chan struct{}),
 	}
+	dataSet := &dataSet{annotations: w.annotations}
 	w.eg.Go(func() error {
-		return w.newWorkerFunc(w.cancelCtx, newPrevChanSet(prev), newNextChanSet(next)).run(byteSet)
+		return w.newWorkerFunc(w.cancelCtx, newPrevChanSet(prev), newNextChanSet(next)).run(dataSet)
 	})
 	w.prev = next
-	w.buf = &bytes.Buffer{}
-	w.annotations = splitAnnotations(w.annotations)
+	lastA := w.annotations[len(w.annotations)-1]
+	_, a := splitAnnotation(lastA, lastA.buf.Len())
+	w.annotations = []*Annotation{a}
+	w.bufSize = 0
 }
 
-// Copy does a cheap copy from a reader to a writer.
-func (w *Writer) Copy(r *Reader, n ...int64) error {
-	c, err := r.ReadCopy(n...)
-	if err != nil {
-		return err
+// Copy copies data from a data reader to the writer.
+// The copy will either be by reading the referenced data, or just
+// copying the data reference (cheap copy).
+func (w *Writer) Copy(dr *DataReader) error {
+	lastA := w.annotations[len(w.annotations)-1]
+	lastA.drs = append(lastA.drs, dr)
+	w.bufSize += int(dr.Len())
+	if w.bufSize > bufSize {
+		w.writeDataSet()
 	}
-	return w.WriteCopy(c)
+	return nil
 }
 
-// WriteCopy writes copy data to the writer.
-func (w *Writer) WriteCopy(c *Copy) error {
-	if _, err := io.Copy(w, c.before); err != nil {
-		return err
+// Close closes the writer.
+func (w *Writer) Close() error {
+	// Write out the last data set.
+	if w.bufSize > 0 {
+		w.writeDataSet()
 	}
-	for _, chunkRef := range c.chunkRefs {
-		w.stats.chunkCount++
-		// (bryce) might want to double check if this is correct.
-		w.stats.annotatedBytesSize += chunkRef.SizeBytes
-		updateAnnotations(chunkRef, nil, w.annotations)
-		if err := w.f(chunkRef, w.annotations); err != nil {
-			return err
+	// Signal to the last worker that it is last.
+	if w.prev != nil {
+		close(w.prev.dataSet)
+		w.prev = nil
+	}
+	// Wait for the workers to finish.
+	return w.eg.Wait()
+}
+
+func joinAnnotations(as []*Annotation, a *Annotation) []*Annotation {
+	// If the annotation being added is the same as the
+	// last, then they are merged.
+	if as != nil {
+		lastA := as[len(as)-1]
+		if lastA.Meta == a.Meta {
+			lastA.buf.Write(a.buf.Bytes())
+			if lastA.tags != nil {
+				lastA.tags = joinTags(lastA.tags, a.tags)
+			}
+			return as
 		}
 	}
-	_, err := io.Copy(w, c.after)
-	return err
+	return append(as, a)
 }
 
-// Close closes the writer.
-func (w *Writer) Close() error {
-	return w.Flush()
+func joinTags(ts1, ts2 []*Tag) []*Tag {
```

Comment:
Similarly could this be called `appendTags`? I'm a little confused why the signature for this is different from `joinAnnotations` above, could you clarify that? Also could the `ts2` be made variadic to match with the builtin `append`.

Reply:
Same as above. It is not just appending. It merges tags at the edges that have the same identifier. The signature is different because that is the use case. I could convert it to a variadic interface, but we are not going to operate on single tags. We join one annotation at a time, and annotations have a set of tags. I think you are trying too hard to pattern match to things you are familiar with, and not considering the subtleties of the use case. This line of thinking is what caused me to spend so much time on the cheap copy. I was too beholden to the notion of `io.Reader` as the only interface for reading from the chunk storage layer and that made cheap copying a nightmare.

Reply:
Same as above, I'm on board with `join` here.